### PR TITLE
Adding ability to set icon_emoji and username on bot messages

### PR DESCRIPTION
**Pull request by EricBartusch**
When using a bot, the iconEmoji and username fields will overwrite the emoji and name used by the bot in Slack.  Either value being set will override the `as_user` field and turn the flag off.  The parameters are optional and not including both of them results in the original behavior.  It's possible to just set one or the other as well.

This works in pipeline and freestyle jobs.

### CODE REVIEW

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 1]
```diff
@@ -390,7 +410,7 @@ public SlackNotifier(final String baseUrl, final String teamDomain, final String
     }
 
     public SlackNotifier(final String baseUrl, final String teamDomain, final String authToken, final boolean botUser, final String room, final String tokenCredentialId,
-                         final String sendAs, final boolean startNotification, final boolean notifyAborted, final boolean notifyFailure,
+                         final String sendAs, final String iconEmoji, final String username, final boolean startNotification, final boolean notifyAborted, final boolean notifyFailure,
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: This is unwieldy, and you would need "another" deprecated SlackNotifier to avoid binary incompatibility.
Perhaps we ought to switch to a builder pattern?
Or is the only used for testing? Perhaps mark it as `NoExternalAccess`
@timja WDYT?

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 2]
```diff
@@ -602,6 +628,20 @@ public void setBotUser(boolean botUser) {
             this.botUser = botUser;
         }
 
+        public String getIconEmoji() { return iconEmoji; }
+
+        @DataBoundSetter
+        public void setIconEmoji(String iconEmoji) {
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Something is fishy when you need to add the setter/getter twice.

Surely we could use the same object/class.

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 3]
```diff
@@ -171,15 +177,26 @@ public boolean publish(String message, JSONArray attachments, String color) {
...
                 if (replyBroadcast) {
                     url += "&reply_broadcast=true";
                 }
                 try {
+                    if (StringUtils.isEmpty(iconEmoji) && StringUtils.isEmpty(username)) {
+                        url += "&as_user=true";
+                    }
+                    else {
+                        if (StringUtils.isNotEmpty(iconEmoji)) {
+                            url += "&icon_emoji=" + URLEncoder.encode(iconEmoji, StandardCharsets.UTF_8.name());
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: May I suggest we start using handy URI?
You should use the Jenkins plugin version of it: https://github.com/jenkinsci/handy-uri-templates-2-api-plugin

However is the docs: https://github.com/damnhandy/Handy-URI-Templates
See usage here: https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/80/files

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 4]
```diff
@@ -231,6 +248,19 @@ private String getTokenToUse(String authTokenCredentialId, String token) {
         return token;
     }
 
+    private String correctEmojiFormat(String iconEmoji) {
+        if(StringUtils.isEmpty(iconEmoji)) {
+            return iconEmoji;
+        }
+        else if (!iconEmoji.startsWith(":")) {
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: You could shorten this to
```suggestion
        iconEmoji = StringUtils.appendIfMissing(iconEmoji, ":")
```

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 5]
```diff
@@ -231,6 +248,19 @@ private String getTokenToUse(String authTokenCredentialId, String token) {
...
+            iconEmoji = ":" + iconEmoji;
+        }
+        if (!iconEmoji.endsWith(":")) {
```
Comment (casz) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: Also
```suggestion
        iconEmoji = StringUtils.prependIfMissing(iconEmoji, ":")
```

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java (hunk 3 above)
Comment (casz) [reply src/main/java/jenkins/plugins/slack/StandardSlackService.java]: See this comment: https://github.com/jenkinsci/slack-plugin/pull/577#discussion_r289592678

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 1 above)
Comment (timja) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: I think we need a Pojo with a builder pattern, we can't keep adding arguments here.

Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Help me understand a little bit: I get the builder pattern bit, but I don't know where it will actually be used.  Wouldn't I need to make another constructor anyway to avoid the binary incompatibility?

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 2 above)
Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Aren't these needed to get/set values on configuration pages?

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 1 above)
Comment (timja) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: There will be one new constructor, and a conversion in the old one to the builder.

Never more will there be a new constructor, there's about 6 in this file

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 2 above)
Comment (casz) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: This could be done as a separate improvement, but there is definitely some need to reduce the repetition :sweat:

File: pom.xml [hunk 6]
```diff
@@ -213,6 +213,11 @@
             <version>1.12</version>
             <optional>true</optional>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.7</version>
```
Comment (casz) [inline pom.xml]: Are we really using such an old commons-lang version? :cry:

**Comment by EricBartusch** [comment]
I'm clueless on what SpotBugs wants from me.  I'm guessing I've implemented the builder pattern incorrectly.

**Comment by timja** [comment]
The error is:
```
[ERROR] Return value of SlackNotifier$SlackNotifierBuilder.build() ignored, but method has no side effect [jenkins.plugins.slack.SlackNotifier] At SlackNotifier.java:[line 452] RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT
```

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 7]
```diff
@@ -389,46 +410,276 @@ public SlackNotifier(final String baseUrl, final String teamDomain, final String
...
     public SlackNotifier(final String baseUrl, final String teamDomain, final String authToken, final boolean botUser, final String room, final String tokenCredentialId,
                          final String sendAs, final boolean startNotification, final boolean notifyAborted, final boolean notifyFailure,
                          final boolean notifyNotBuilt, final boolean notifySuccess, final boolean notifyUnstable, final boolean notifyRegression, final boolean notifyBackToNormal,
                          final boolean notifyRepeatedFailure, final boolean includeTestSummary, final boolean includeFailedTests, MatrixTriggerMode matrixTriggerMode,
                          CommitInfoChoice commitInfoChoice, boolean includeCustomMessage, String customMessage, String customMessageSuccess,
                          String customMessageAborted, String customMessageNotBuilt, String customMessageUnstable, String customMessageFailure) {
-        super();
-        this.baseUrl = baseUrl;
+
+        new SlackNotifier.SlackNotifierBuilder()
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: this is the spotbugs issue, you're building a new object but not assigning it anywhere

Comment (timja) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: you should be calling this constructor instead with your builder:
```
private SlackNotifier(SlackNotifierBuilder slackNotifierBuilder) {
```

**Comment by timja** [comment]
I've pushed a commit to cleanup a couple of small issues.
I'll do a bit more of a review and then test it out locally.

Thanks for the PR!

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 8]
```diff
@@ -42,54 +39,68 @@
...
      * @param botUser
      * @param roomId         a semicolon separated list of rooms to notify
      * @param replyBroadcast
+     * @param iconEmoji
+     * @param username
      * @param populatedToken a non-null token to use for authentication
      */
-    public StandardSlackService(String baseUrl, String teamDomain, boolean botUser, String roomId, boolean replyBroadcast, String populatedToken) {
-        this(baseUrl, teamDomain, botUser, roomId, replyBroadcast);
+    public StandardSlackService(String baseUrl, String teamDomain, boolean botUser, String roomId, boolean replyBroadcast, String iconEmoji, String username, String populatedToken) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: this also needs a builder...

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 9]
```diff
@@ -231,6 +248,16 @@ private String getTokenToUse(String authTokenCredentialId, String token) {
         return token;
     }
 
+    private String correctEmojiFormat(String iconEmoji) {
+        if(StringUtils.isEmpty(iconEmoji)) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]:
```suggestion
        if (StringUtils.isEmpty(iconEmoji)) {
```

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java (hunk 8 above)
Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/StandardSlackService.java]: Now that I have a working example, I should be able to do a better job setting one up.

File: src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/config.jelly [hunk 10]
```diff
@@ -20,6 +20,12 @@
         <f:entry field="botUser" title="Is Bot User?">
             <f:checkbox />
         </f:entry>
+            <f:entry field="iconEmoji" title="Icon Emoji">
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/config.jelly]: formatting looks off here

File: src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-botUser.html [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<div>
+    <p>Bot user option indicates the token belongs to a bot user in Slack.</p>
+    <p>If the notification will be sent to a user via direct message, the default integration sends it via @slackbot, use this option if you want to send messages via a bot user.</p>
+</div>
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-botUser.html]:
```suggestion
</div>

```

File: src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-iconEmoji.html [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom emoji to use as the bot's icon in Slack</p>
+</div>
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-iconEmoji.html]:
```suggestion
</div>

```

File: src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-iconEmoji.html [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom emoji to use as the bot's icon in Slack</p>
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-iconEmoji.html]:
```suggestion
    <p>Choose a custom emoji to use as the bot's icon in Slack, requires using a bot user</p>
```

File: src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-username.html [hunk 14]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom username to use as the bot's name in Slack</p>
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/workflow/SlackSendStep/help-username.html]:
```suggestion
    <p>Choose a custom username to use as the bot's name, requires using a bot user</p>
```

File: src/main/webapp/help-globalConfig-iconEmoji.html [hunk 15]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom emoji to use as the bot's icon in Slack</p>
```
Comment (timja) [inline src/main/webapp/help-globalConfig-iconEmoji.html]:
```suggestion
    <p>Choose a custom emoji to use as the bot's icon in Slack, requires using a bot user</p>
```

File: src/main/webapp/help-globalConfig-username.html [hunk 16]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom username to use as the bot's name in Slack</p>
```
Comment (timja) [inline src/main/webapp/help-globalConfig-username.html]:
```suggestion
    <p>Choose a custom username to use as the bot's name in Slack, requires using a bot user</p>
```

File: src/main/webapp/help-projectConfig-iconEmoji.html [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom emoji to use as the bot's icon in Slack</p>
```
Comment (timja) [inline src/main/webapp/help-projectConfig-iconEmoji.html]:
```suggestion
    <p>Choose a custom emoji to use as the bot's icon in Slack, requires using a bot user</p>
```

File: src/main/webapp/help-projectConfig-username.html [hunk 18]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Works only when sending with a bot user. Choose a custom username to use as the bot's name in Slack</p>
```
Comment (timja) [inline src/main/webapp/help-projectConfig-username.html]:
```suggestion
    <p>Choose a custom username to use as the bot's name in Slack, requires using a bot user</p>
```

File: src/test/java/jenkins/plugins/slack/SlackNotifierStub.java [hunk 19]
```diff
@@ -11,12 +11,38 @@ public SlackNotifierStub(String baseUrl, String teamDomain, String authToken, bo
...
+                .withCustomMessage(customMessage)
+                .withCustomMessageSuccess(customMessageSuccess)
+                .withCustomMessageAborted(customMessageAborted)
+                .withCustomMessageNotBuilt(customMessageNotBuilt)
+                .withCustomMessageUnstable(customMessageUnstable)
+                .withCustomMessageFailure(customMessageFailure)
+        );
     }
 
+
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/SlackNotifierStub.java]:
```suggestion
```

File: src/test/java/jenkins/plugins/slack/StandardSlackServiceTest.java [hunk 20]
```diff
@@ -13,7 +13,7 @@
      */
     @Test
     public void publishWithBadHostShouldNotRethrowExceptions() {
-        StandardSlackService service = new StandardSlackService("", "foo", false, "#general", false, "token");
+        StandardSlackService service = new StandardSlackService("", "foo", false, "#general", false,"token");
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/StandardSlackServiceTest.java]:
```suggestion
        StandardSlackService service = new StandardSlackService("", "foo", false, "#general", false, "token");
```

File: src/test/java/jenkins/plugins/slack/StandardSlackServiceTest.java [hunk 21]
```diff
@@ -23,7 +23,7 @@ public void publishWithBadHostShouldNotRethrowExceptions() {
      */
     @Test
     public void invalidTeamDomainShouldFail() {
-        StandardSlackService service = new StandardSlackService("", "my", false, "#general", false, "token");
+        StandardSlackService service = new StandardSlackService("", "my", false, "#general", false,"token");
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/StandardSlackServiceTest.java]:
```suggestion
        StandardSlackService service = new StandardSlackService("", "my", false, "#general", false, "token");
```

File: src/test/java/jenkins/plugins/slack/StandardSlackServiceTest.java [hunk 22]
```diff
@@ -32,7 +32,7 @@ public void invalidTeamDomainShouldFail() {
      */
     @Test
     public void invalidTokenShouldFail() {
-        StandardSlackService service = new StandardSlackService("", "tinyspeck", false, "#general", false, "token");
+        StandardSlackService service = new StandardSlackService("", "tinyspeck", false, "#general", false,"token");
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/StandardSlackServiceTest.java]:
```suggestion
        StandardSlackService service = new StandardSlackService("", "tinyspeck", false, "#general", false, "token");
```

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 23]
```diff
@@ -671,7 +757,14 @@ SlackService getSlackService(final String baseUrl, final String teamDomain, fina
         SlackService getSlackService(final String baseUrl, final String teamDomain, final String authTokenCredentialId, final boolean botUser, final String roomId, final Item item) {
             final String populatedToken = CredentialsObtainer.getTokenToUse(authTokenCredentialId, item,null );
             if (populatedToken != null) {
-                return new StandardSlackService(baseUrl, teamDomain, botUser, roomId, false, populatedToken);
+                return new StandardSlackService(
+                        new StandardSlackServiceBuilder()
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]:
```suggestion
                        StandardSlackService.builder()
```

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 24]
```diff
@@ -66,42 +65,48 @@ public StandardSlackService(String baseUrl, String teamDomain, String token, Str
...
-     * @param teamDomain     the teamDomain inside slack.com to use
-     * @param botUser
-     * @param roomId         a semicolon separated list of rooms to notify
-     * @param replyBroadcast
-     * @param populatedToken a non-null token to use for authentication
-     */
+    @Deprecated
     public StandardSlackService(String baseUrl, String teamDomain, boolean botUser, String roomId, boolean replyBroadcast, String populatedToken) {
-        this(baseUrl, teamDomain, botUser, roomId, replyBroadcast);
+        this(new StandardSlackServiceBuilder()
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: sorry I missed this before
```suggestion
        this(builder()
```

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 25]
```diff
@@ -151,49 +156,53 @@ public boolean publish(String message, JSONArray attachments, String color) {
...
+                        json.put("username", username);
                     }
-                    url += "&attachments=" + URLEncoder.encode(attachments.toString(), StandardCharsets.UTF_8.name());
-                } catch (UnsupportedEncodingException e) {
-                    logger.log(Level.ALL, "Error while encoding payload: " + e.getMessage());
                 }
-                post = new HttpPost(url);
             }
+
+            nvps.add(new BasicNameValuePair("payload", json.toString()));
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java]: this doesn't appear to be used any more?

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 26]
```diff
@@ -671,7 +757,14 @@ SlackService getSlackService(final String baseUrl, final String teamDomain, fina
         SlackService getSlackService(final String baseUrl, final String teamDomain, final String authTokenCredentialId, final boolean botUser, final String roomId, final Item item) {
             final String populatedToken = CredentialsObtainer.getTokenToUse(authTokenCredentialId, item,null );
             if (populatedToken != null) {
-                return new StandardSlackService(baseUrl, teamDomain, botUser, roomId, false, populatedToken);
+                return new StandardSlackService(
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: you aren't passing through icon emoji or username here, I was testing the new fields out in the global configuration and it didn't work...

Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Crap, must have broken with the builder change. I'll take a look.

Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: I lied, this was never working.  I was using empty string as a default value when iconEmoji and username weren't set instead of what was in the global config.  This is set in SlackSendStep.java.

**Comment by timja** [comment]
You need to update `doTestConnection` in SlackNotifier
So that the "Test connection" works on the global configuration page

**Comment by EricBartusch** [comment]
Can I also change the name of the method and button? I didn't think to add it because `doTestConnection` makes our dinner like it's just making sure Jenkins can talk to Slack, not actually send a message. `doTestMessage` or `sendTestMessage` makes more sense to me.

**Comment by timja** [comment]
it needs to start with `do` it makes sense to me how it is, it's testing the connection to slack...

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 27]
```diff
@@ -712,12 +815,18 @@ public FormValidation doTestConnection(@QueryParameter("baseUrl") final String b
                     targetUrl = this.baseUrl;
                 }
                 String targetDomain = Util.fixEmpty(teamDomain) != null ? teamDomain : this.teamDomain;
-                boolean targetBotUser = botUser || this.botUser;
-                String targetTokenCredentialId = Util.fixEmpty(tokenCredentialId) != null ? tokenCredentialId :
-                        this.tokenCredentialId;
                 String targetRoom = Util.fixEmpty(room) != null ? room : this.room;
 
-                SlackService testSlackService = getSlackService(targetUrl, targetDomain, targetTokenCredentialId, targetBotUser, targetRoom, project);
+                SlackService testSlackService = getSlackService(new StandardSlackServiceBuilder()
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]:
```suggestion
                SlackService testSlackService = getSlackService(StandardSlackServiceBuilder.builder()
```

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 28]
```diff
@@ -712,12 +815,18 @@ public FormValidation doTestConnection(@QueryParameter("baseUrl") final String b
...
+                        .withBaseUrl(targetUrl)
+                        .withTeamDomain(targetDomain)
+                        .withBotUser(botUser)
+                        .withRoomId(targetRoom)
+                        .withIconEmoji(iconEmoji)
+                        .withUsername(username)
+                        ,tokenCredentialId, project
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: what are these doing here?

**Comment by timja** [comment]
Test connection is working in global config now:
image

**Comment by timja** [comment]
The slack send pipeline log hasn't been updated:
image

but the functionality works

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 29]
```diff
@@ -712,12 +815,18 @@ public FormValidation doTestConnection(@QueryParameter("baseUrl") final String b
                     targetUrl = this.baseUrl;
                 }
                 String targetDomain = Util.fixEmpty(teamDomain) != null ? teamDomain : this.teamDomain;
-                boolean targetBotUser = botUser || this.botUser;
-                String targetTokenCredentialId = Util.fixEmpty(tokenCredentialId) != null ? tokenCredentialId :
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: removing these two lines break relying on global config when testing in a freestyle job

Comment (EricBartusch) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: You're right, the test needs to be consistent so the behavior is the same between the test and how it actually works.

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 30]
```diff
@@ -471,7 +532,16 @@ public SlackService newSlackService(AbstractBuild abstractBuild, BuildListener l
         authTokenCredentialId = env.expand(authTokenCredentialId);
         room = env.expand(room);
         final String populatedToken = CredentialsObtainer.getTokenToUse(authTokenCredentialId, abstractBuild.getParent(), authToken);
-        return new StandardSlackService(baseUrl, teamDomain, botUser, room, false, populatedToken);
+        return new StandardSlackService(
+                new StandardSlackServiceBuilder()
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]:
```suggestion
                builder()
```

File: src/main/java/jenkins/plugins/slack/workflow/SlackSendStep.java [hunk 31]
```diff
@@ -363,8 +386,18 @@ private String defaultIfEmpty(String value) {
         }
 
         //streamline unit testing
-        SlackService getSlackService(String baseUrl, String team, boolean botUser, String channel, boolean replyBroadcast, String populatedToken) {
-            return new StandardSlackService(baseUrl, team, botUser, channel, replyBroadcast, populatedToken);
+        SlackService getSlackService(String baseUrl, String team, boolean botUser, String channel, boolean replyBroadcast, String iconEmoji, String username, String populatedToken) {
+            return new StandardSlackService(
+                    new StandardSlackServiceBuilder()
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/workflow/SlackSendStep.java]:
```suggestion
                    StandardSlackService.builder()
```

File: src/test/java/jenkins/plugins/slack/SlackNotifierStub.java [hunk 32]
```diff
@@ -11,10 +11,35 @@ public SlackNotifierStub(String baseUrl, String teamDomain, String authToken, bo
                              boolean notifyRepeatedFailure, boolean includeTestSummary, boolean includeFailedTests, MatrixTriggerMode matrixTriggerMode,
                              CommitInfoChoice commitInfoChoice, boolean includeCustomMessage, String customMessage, String customMessageSuccess,
                              String customMessageAborted, String customMessageNotBuilt, String customMessageUnstable, String customMessageFailure) {
-        super(baseUrl, teamDomain, authToken, botUser, room, authTokenCredentialId, sendAs, startNotification, notifyAborted, notifyFailure,
-                notifyNotBuilt, notifySuccess, notifyUnstable, notifyRegression, notifyBackToNormal, notifyRepeatedFailure,
-                includeTestSummary, includeFailedTests, matrixTriggerMode, commitInfoChoice, includeCustomMessage, customMessage,
-                customMessageSuccess, customMessageAborted, customMessageNotBuilt, customMessageUnstable, customMessageFailure);
+        super(new SlackNotifierBuilder()
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/SlackNotifierStub.java]:
```suggestion
        super(builder()
```

**Comment by timja** [comment]
Can you update the tests to not call the deprecated constructor please?
i.e. in:
StandardSlackServiceTest
StandardSlackServiceStub
SlackNotifierStub (this one maybe needs a suppression, haven't looked closely)

**Comment by timja** [comment]
Want me to wait for your other PR before doing a release?

**Comment by EricBartusch** [comment]
Yeah, I should be able to do the other one pretty soon here
